include/dgl/lazy.h

```diff
@@ -7,14 +7,38 @@
 #define DGL_LAZY_H_
 
 #include <memory>
+#include <mutex>
+#include <atomic>
 
 namespace dgl {
 
+/*!
+ * \brief Lazy initialization
+ */
+template <typename T, typename Fn>
+const std::shared_ptr<T> LazyInit(std::shared_ptr<T>* ptr, std::mutex *mutex, Fn func) {
+  std::shared_ptr<T> result;
+
+  while (!(result = std::atomic_load_explicit(ptr, std::memory_order_acquire))) {
+    std::lock_guard<std::mutex> lock(*mutex);
+    result = *ptr;
+
```

Comment:
For this kind of check and lock, you should do check, lock and check style.
Otherwise, there is still data race here.

Reply:
Line 22 does the first check.
Line 23 locks.
Line 26 does the second check.
Correct me if I'm wrong.